truffleruby/src/main/java/org/truffleruby/core/binding/BindingNodes.java

```diff
@@ -373,28 +389,37 @@ protected int getCacheLimit() {
 
         @Specialization
         public DynamicObject localVariables(DynamicObject binding) {
+            MaterializedFrame extras = Layouts.BINDING.getExtras(binding);
             MaterializedFrame frame = Layouts.BINDING.getFrame(binding);
 
-            return listLocalVariables(getContext(), frame);
+            return listLocalVariables(getContext(), frame, extras);
         }
 
         @TruffleBoundary
-        public static DynamicObject listLocalVariables(RubyContext context, Frame frame) {
+        public static DynamicObject listLocalVariables(RubyContext context, Frame frame, Frame extras) {
             final Set<Object> names = new LinkedHashSet<>();
+            if (extras != null) {
+                addNamesFromFrame(context, extras, names);
+            }
             while (frame != null) {
-                for (FrameSlot slot : frame.getFrameDescriptor().getSlots()) {
-                    if (slot.getIdentifier() instanceof String &&
-                            !((String) slot.getIdentifier()).startsWith("rubytruffle_temp_frame_on_stack_marker") &&
-                            !Translator.FRAME_LOCAL_GLOBAL_VARIABLES.contains(slot.getIdentifier())) {
-                        names.add(context.getSymbolTable().getSymbol((String) slot.getIdentifier()));
-                    }
-                }
+                addNamesFromFrame(context, frame, names);
 
                 frame = RubyArguments.getDeclarationFrame(frame);
             }
```

Comment:
This could be unified nicely if the extra's declaration frame is the captured frame.